lib/kinesis_client.rb

```diff
@@ -7,42 +7,96 @@
 class KinesisClient
   attr_reader :config, :avro
 
-  def initialize(config)
+  def initialize(config, batch_size = 1, automatically_push = true)
     @config = config
+    @stream_name = @config[:stream_name]
     @avro = nil
+    @batch_size = batch_size
+    @batch = []
+    @automatically_push = automatically_push
+    @client_options = config[:profile] ? { profile: config[:profile] } : {}
+    @client = Aws::Kinesis::Client.new @client_options
 
     if config[:schema_string]
       @avro = NYPLAvro.by_name(config[:schema_string])
     end
 
+    @shovel_method = batch_size > 1 ? :push_to_batch : :push_record
+
   end
 
-  def <<(json_message)
+  def convert_to_record(json_message)
     if config[:schema_string]
       message = avro.encode(json_message, false)
     else
       message = json_message
     end
 
-    client = Aws::Kinesis::Client.new
     partition_key = (config[:partition_key] ? json_message[config[:partition_key]] : SecureRandom.hex(20)).hash.to_s
-
-    resp = client.put_record({
-      stream_name: config[:stream_name],
+    {
       data: message,
       partition_key: partition_key
-      })
-
-      return_hash = {}
-
-      if resp.successful?
-        return_hash["code"] = "200"
-        return_hash["message"] = json_message, resp
-        $logger.info("Message sent to #{config[:stream_name]} #{json_message}, #{resp}") if $logger
-      else
-        $logger.error("message" => "FAILED to send message to HoldRequestResult #{json_message}, #{resp}.") if $logger
-        raise NYPLError.new json_message, resp
-      end
-      return_hash
+    }
+
+  end
+
+  def <<(json_message)
+    send(@shovel_method, json_message)
+  end
+
+  def push_record(json_message)
```

Comment:
Could this be defined as `push_batch([json_message])` now?

Reply:
I was wondering about that. I think we probably could go that way, but the error handling is a little different for the two cases. When we push a single record, it raises an error if the record fails. When we push a batch, it just records the number of failures in the return value. Maybe we don't care about that. Also, it would in principle be a breaking change if we did that, although I don't know if it would break anything in practice.

Reply:
Yeah fair enough